### PR TITLE
S3 layer provider

### PR DESCRIPTION
**Pull request by jamie-harness**
S3 Cache provider change

functional working

### CODE REVIEW

File: cmd/buildah/build.go [hunk 1]
```diff
@@ -162,6 +167,20 @@ func buildCmd(c *cobra.Command, inputArgs []string, iopts buildOptions) error {
 		}
 	}
 
+	if c.Flag("s3-local-cache-dir").Value.String() != "" {
```
Comment (george-harness) [inline cmd/buildah/build.go:171]: please add check that file-cache-dir and s3-local-cache-dir cannot be specified at the same time. There are some examples of such checks

File: imagebuildah/cache/file_transport/file_transport.go [hunk 2]
```diff
@@ -6,6 +6,8 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/containers/buildah/define"
+
```
Comment (george-harness) [inline imagebuildah/cache/file_transport/file_transport.go]: please remove the empty line

File: imagebuildah/cache/s3_layer_provider.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+		if destroyErr := policyContext.Destroy(); destroyErr != nil {
+			if err == nil {
+				err = destroyErr
+			} else {
+				err = errors.Wrap(err, destroyErr.Error())
+			}
+		}
+	}()
+
+	imageIDBytes, err := ioutil.ReadFile(slp.imageIDFilepath(layerKey))
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go:95]: please check for the err

File: imagebuildah/cache/s3_layer_provider.go [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+	imageID := string(imageIDBytes)
+
+	destRef, err := is.Transport.ParseStoreReference(slp.store, "@"+imageID)
+	if err != nil {
+		return "", err
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go]: please wrap the error

File: imagebuildah/cache/s3_layer_provider.go [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+	_, err = copy.Image(ctx, policyContext, destRef, srcRef, nil)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to obtain the image for %s", layerKey)
+	}
+
+	return imageID, nil
+}
+
+func (slp *S3LayerProvider) tryDownloadLayerFromS3(layerKey string, dir string) bool {
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go]: please separate the result from the error. return true, nil id succeed, false, nil if missing and if operation fail return false, err

File: imagebuildah/cache/s3_layer_provider.go [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+		return false
+	}
+	content, err := os.ReadFile(path.Join(dir, "imageID"))
+	if err != nil {
+		return false
+	}
+	imageId := string(content)
+	fpath := path.Join(slp.location, "blobs", imageId)
+	if err := os.MkdirAll(filepath.Dir(fpath), 0775); err != nil {
+		panic(err)
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go]: please remove any panic from the code, prints also

File: cmd/buildah/build.go (hunk 1 above)
Comment (jamie-harness) [reply cmd/buildah/build.go:171]: I was not able to find some examples of this. I'm just making an explicit if check now. Please let me know if there's a better way. Thank you!

File: imagebuildah/cache/file_transport/file_transport_dest.go [hunk 7]
```diff
@@ -5,9 +5,17 @@ import (
 	"io"
 	"io/ioutil"
 	"os"
+	"path"
 	"path/filepath"
 	"runtime"
 
+	"github.com/aws/aws-sdk-go/service/s3"
+
```
Comment (george-harness) [inline imagebuildah/cache/file_transport/file_transport_dest.go]: please remove the empty line

File: imagebuildah/cache/file_transport/file_transport_dest.go [hunk 8]
```diff
@@ -5,9 +5,17 @@ import (
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+
```
Comment (george-harness) [inline imagebuildah/cache/file_transport/file_transport_dest.go]: ditto

File: imagebuildah/cache/file_transport/file_transport_dest.go [hunk 9]
```diff
@@ -186,6 +195,35 @@ func (d *fileImageDestination) PutBlob(ctx context.Context, stream io.Reader, in
...
+	if d.ref.s3Options != nil {
+		blobToUpload, err := os.Open(blobFile.Name())
+		if err != nil {
+			return types.BlobInfo{}, err
+		}
+		s3Config := &aws.Config{
+			Credentials:      credentials.NewStaticCredentials(d.ref.s3Options.S3Key, d.ref.s3Options.S3Secret, ""),
+			Endpoint:         aws.String(d.ref.s3Options.S3EndPoint),
+			Region:           aws.String(d.ref.s3Options.S3Region),
+			DisableSSL:       aws.Bool(true),
```
Comment (george-harness) [inline imagebuildah/cache/file_transport/file_transport_dest.go]: that likely should be a flag as well

File: imagebuildah/cache/s3_layer_provider.go [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+// PopulateLayer scans the local images and adds them to the map.
+func (slp *S3LayerProvider) PopulateLayer(ctx context.Context, topLayer string) error {
+	return nil
+}
+
+// Load returns the image id for the key.
+func (slp *S3LayerProvider) Load(ctx context.Context, layerKey string) (string, error) {
+	dir := slp.keyDirectory(layerKey)
+
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go:64]: this check, might return false positive if the download was interrupted and left the directory in a bad state or even empty. We need to improve this

we can delete the directory if the download failed

We also need to make sure if this is multi-thread and multi-process safe

File: imagebuildah/cache/s3_layer_provider.go [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+	}
+
+	srcRef, err := file_transport.NewReference(dir, slp.s3CacheOptions)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create reference for %s", layerKey)
+	}
+
+	policyContext, err := util.GetPolicyContext(slp.systemContext)
+	if err != nil {
+		return "", err
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go]: please wrap this error

File: imagebuildah/cache/s3_layer_provider.go [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+	d := downloader{bucket: slp.s3CacheOptions.S3Bucket, dir: slp.location, Downloader: manager}
+	input := &s3.ListObjectsInput{
+		Bucket: aws.String(slp.s3CacheOptions.S3Bucket),
+		Prefix: aws.String(layerKey),
+	}
+	client := s3.New(sess)
+	err := client.ListObjectsPages(input, d.eachPage)
+	if err != nil {
+		return false, err
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go]: ditto

File: imagebuildah/cache/s3_layer_provider.go [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+	}
+	content, err := os.ReadFile(path.Join(dir, "imageID"))
+	if err != nil {
+		// cache does not exist
+		return false, nil
+	}
+	imageId := string(content)
+	fpath := path.Join(slp.location, "blobs", imageId)
+	if slp.downloadFileIfNotExist(fpath, path.Join("blobs", imageId), manager) != nil {
+		return false, err
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go]: ditto

File: imagebuildah/cache/s3_layer_provider.go [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+		return false, errors.Wrap(err, "Manifest is not a valid Json")
+	}
+	for _, layer := range manifest.Layers {
+		if len(layer.Digest) < 7 {
+			return false, errors.Wrapf(err, "Failed to read manifest from downloaded cache for layer %s", layerKey)
+		}
+		imageId = layer.Digest[7:]
+		fpath := path.Join(slp.location, "blobs", imageId)
+		if slp.downloadFileIfNotExist(fpath, path.Join("blobs", imageId), manager) != nil {
+			return false, err
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go]: ditto

File: imagebuildah/cache/s3_layer_provider.go [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+		}
+	}
+	return true, nil
+}
+
+func (slp *S3LayerProvider) downloadFileIfNotExist(fpath string, s3path string, manager *s3manager.Downloader) error {
+	if _, err := os.Stat(fpath); os.IsNotExist(err) {
```
Comment (george-harness) [inline imagebuildah/cache/s3_layer_provider.go]: ditto

File: imagebuildah/cache/s3_layer_provider.go (hunk 13 above)
Comment (jamie-harness) [reply imagebuildah/cache/s3_layer_provider.go]: this one is wrapped inside downloadFileIfNotExist function

File: imagebuildah/cache/s3_layer_provider.go (hunk 14 above)
Comment (jamie-harness) [reply imagebuildah/cache/s3_layer_provider.go]: this one is wrapped inside downloadFileIfNotExist function

Comment (george-harness) [reply imagebuildah/cache/s3_layer_provider.go]: go does not provide call stack when reporting errors. This makes it extremely hard to track what is going on. The best way to track errors is with providing relatively unique message that can identify every call you went through

File: imagebuildah/cache/s3_layer_provider.go (hunk 10 above)
Comment (jamie-harness) [reply imagebuildah/cache/s3_layer_provider.go:64]: added todo

File: imagebuildah/cache/s3_layer_provider.go [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+		// image with that key is not in the cache
+		// TODO: safer download if the program get interrupted
+		existInS3, err := slp.tryDownloadLayerFromS3(layerKey, dir)
+		if !existInS3 {
+			return "", nil
+		}
+		if err != nil {
+			logrus.Warnf("Attempt to download from S3 Cache failed due to: %s", err.Error())
+			// the normal build without cache should be performed
+			return "", nil
```
Comment (jamie-harness) [inline imagebuildah/cache/s3_layer_provider.go]: should return err

**Pull request merged**